components/helm-broker/internal/controller/addonsconfiguration_controller.go

```diff
@@ -0,0 +1,500 @@
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+		updatedInstance, err := r.addFinalizer(pendingInstance)
+		if err != nil {
+			return reconcile.Result{Requeue: true}, exerr.Wrapf(err, "while adding a finalizer to AddonsConfiguration %q", request.NamespacedName)
+		}
+		err = r.addAddonsProcess(updatedInstance)
+		if err != nil {
+			return reconcile.Result{}, exerr.Wrapf(err, "while creating AddonsConfiguration %q", request.NamespacedName)
+		}
+		r.log.Info("Add AddonsConfiguration process completed")
+
+	} else if instance.Generation > instance.Status.ObservedGeneration {
+		r.log.Infof("Start update AddonsConfiguration %s/%s process", instance.Name, instance.Namespace)
+
+		instance.Status = addonsv1alpha1.AddonsConfigurationStatus{}
+		err = r.addAddonsProcess(instance)
+		if err != nil {
+			return reconcile.Result{}, exerr.Wrapf(err, "while updating AddonsConfiguration %q", request.NamespacedName)
+		}
+		r.log.Info("Update AddonsConfiguration process completed")
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func (r *ReconcileAddonsConfiguration) addAddonsProcess(addon *addonsv1alpha1.AddonsConfiguration) error {
+	repositories := addons.NewRepositoryCollection()
+
+	r.log.Infof("- load bundles and charts for each addon")
+	for _, specRepository := range addon.Spec.Repositories {
+		if err := specRepository.VerifyURL(r.developMode); err != nil {
+			r.log.Errorf("url %q address is not valid: %s", specRepository.URL, err)
+			continue
+		}
+		r.log.Infof("- create addons for %q repository", specRepository.URL)
+		repo := addons.NewAddonsRepository(specRepository.URL)
+
+		addons, err := r.createAddons(specRepository.URL)
+		if err != nil {
+			repo.Failed()
+			repo.Repository.Reason = addonsv1alpha1.RepositoryURLFetchingError
+			repo.Repository.Message = err.Error()
+			repositories.AddRepository(repo)
+
+			r.log.Errorf("while creating addons for repository from %q: %s", specRepository.URL, err)
+			continue
+		}
+
+		repo.Addons = addons
+		repositories.AddRepository(repo)
+	}
+
+	r.log.Info("- check duplicate ID addons alongside repositories")
+	repositories.ReviseBundleDuplicationInRepository()
+
+	r.log.Info("- check duplicates ID addons in existing addons configuration")
+	list, err := r.existingAddonsConfigurationList(addon)
+	if err != nil {
+		r.log.Errorf("cannot fetch AddonsConfiguration list: %s", err)
+		return exerr.Wrap(err, "while fetching addons configuration list")
+	}
+	repositories.ReviseBundleDuplicationInStorage(list)
+
+	r.log.Info("- save ready bundles and charts in storage")
+	r.saveBundle(internal.Namespace(addon.Namespace), repositories)
+
+	r.log.Info("- update AddonsConfiguration status")
+	r.statusSnapshot(addon, repositories)
+	if err = r.updateAddonStatus(addon); err != nil {
+		r.log.Errorf("cannot update AddonsConfiguration %s: %v", addon.Name, err)
+		return exerr.Wrap(err, "while update AddonsConfiguration status")
+	}
+
+	if err := r.deleteOrphanBundles(addon.Namespace); err != nil {
```

Comment:
consult this method with @piotrmiskiewicz

Reply:
done